Sources/HomomorphicEncryption/PolyRq/PolyRq.swift

```diff
@@ -408,21 +408,42 @@ extension PolyRq where F == Coeff {
 
 extension PolyRq where F == Coeff {
     @inlinable
-    mutating func multiplyInversePowerOfX(_ power: Int) throws {
-        precondition(power >= 0)
-        let effectiveStep = power % (degree &<< 1)
-        if effectiveStep == 0 {
-            return
+    mutating func multiplyPowerOfX(_ power: Int) throws {
+        // Calculate effective step once, handling both positive and negative powers
+        let twiceDegree = degree &<< 1
+        let absEffectiveStep = abs(power) % twiceDegree
+
+        if absEffectiveStep == 0 {
+            return // No change needed for powers that are multiples of 2*degree
         }
-        try data.rotate(range: degree, step: effectiveStep)
+
+        // Determine rotation direction and effective step based on power sign
+        let rotationStep = power < 0 ? -absEffectiveStep : absEffectiveStep
+        try data.rotateColumns(by: rotationStep % degree)
+
+        // Negate appropriate coefficients based on power sign
         for (rowIndex, modulus) in moduli.enumerated() {
-            if effectiveStep < degree {
-                for columnIndex in degree &- effectiveStep..<degree {
-                    data[rowIndex, columnIndex] = data[rowIndex, columnIndex].negateMod(modulus: modulus)
+            if power < 0 {
+                // Negative power case
+                if absEffectiveStep < degree {
```

Comment:
These if statements for power and effective steps appear to be invariant inside the for loop. Move them outside and do the branching only once. You get the minor inconvenience that the outer for loop line is duplicated. However, intent is clear and at least in C++ cache lines would be better utilised and thus more performant.

Reply:
Having a second look, I noticed to that the only thing that is changing in the different for loops is the range object that is created to iterate through. If we create the object first based on the branching decision, then you only need the one general for loop.